### PR TITLE
P2p gossip propagate once

### PR DESCRIPTION
**Pull request by y0sher**

### CODE REVIEW

File: p2p/gossip/protocol.go [hunk 1]
```diff
@@ -77,11 +76,11 @@ func NewProtocol(config config.SwarmConfig, base baseNetwork, localNodePubkey p2
 		peers:           make(map[string]*peer),
 		shutdown:        make(chan struct{}),
 		oldMessageQ:     make(map[hash]struct{}), // todo : remember to drain this
-		invalidMessageQ: make(map[hash]bool),     // todo : remember to drain this
-		peersMutex:      sync.RWMutex{},
-		relayQ:          relayChan,
-		messageQ:        make(chan protocolMessage, messageQBufferSize),
-		propagateQ:      make(chan service.MessageValidation, propagateHandleBufferSize),
+		//invalidMessageQ: make(map[hash]bool),     // todo : remember to drain this
```
Comment (y0sher) [inline p2p/gossip/protocol.go]: to remove

File: p2p/gossip/protocol.go [hunk 2]
```diff
@@ -93,48 +92,32 @@ type sender interface {
...
+			p.Log.Info("Gossip protocol failed to send msg (calcHash %d) to peer %v, second attempt. err=%v", checksum, p.pubkey, err)
+			return
 		}
-		p.msgMutex.Lock()
-		p.knownMessages[checksum] = struct{}{}
-		p.msgMutex.Unlock()
-	}()
-	return nil
+	}
+	return
```
Comment (y0sher) [inline p2p/gossip/protocol.go]: not needed

File: p2p/gossip/protocol.go [hunk 3]
```diff
@@ -93,48 +92,32 @@ type sender interface {
...
-	}
-	p.msgMutex.RUnlock()
-	go func() {
-		p.Debug("sending message to peer %v, hash %d", p.pubkey, checksum)
-		err := p.net.SendMessage(p.pubkey, ProtocolName, msg)
+func (p *peer) send(msg []byte, checksum hash) {
+	p.Debug("sending message to peer %v, hash %d", p.pubkey, checksum)
+	err := p.net.SendMessage(p.pubkey, ProtocolName, msg)
+	if err != nil {
+		p.Log.Info("Gossip protocol failed to send msg (calcHash %d) to peer %v, first attempt. err=%v", checksum, p.pubkey, err)
```
Comment (zalmen) [inline p2p/gossip/protocol.go]: This should be an Error, not info.

Comment (zalmen) [reply p2p/gossip/protocol.go]: (I know it's my code :))

File: p2p/gossip/protocol.go [hunk 4]
```diff
@@ -93,48 +92,32 @@ type sender interface {
...
+		err = p.net.SendMessage(p.pubkey, "", msg)
 		if err != nil {
-			p.Log.Info("Gossip protocol failed to send msg (calcHash %d) to peer %v, first attempt. err=%v", checksum, p.pubkey, err)
-			// doing one retry before giving up
-			err = p.net.SendMessage(p.pubkey, "", msg)
-			if err != nil {
-				p.Log.Info("Gossip protocol failed to send msg (calcHash %d) to peer %v, second attempt. err=%v", checksum, p.pubkey, err)
-				return
-			}
+			p.Log.Info("Gossip protocol failed to send msg (calcHash %d) to peer %v, second attempt. err=%v", checksum, p.pubkey, err)
```
Comment (zalmen) [inline p2p/gossip/protocol.go]: also should be an error

File: p2p/gossip/protocol.go [hunk 5]
```diff
@@ -206,9 +160,14 @@ func (prot *Protocol) propagateMessage(payload []byte, h hash, nextProt string)
 	}
 
 	prot.peersMutex.RLock()
+peerLoop:
 	for p := range prot.peers {
-		peer := prot.peers[p]
-		peer.send(data, h) // non blocking
+		for e := range exclude {
```
Comment (zalmen) [inline p2p/gossip/protocol.go]: In which case do you expect seeing more than one sender?

Comment (zalmen) [reply p2p/gossip/protocol.go]: Currently, the only call to this method passes one sender

File: p2p/gossip/protocol.go [hunk 6]
```diff
@@ -229,16 +188,15 @@ func (prot *Protocol) Broadcast(payload []byte, nextProt string) error {
...
 	}
 
-	fmt.Println("PAYLOAD : ", payload)
 	msg := &pb.ProtocolMessage{
 		Metadata: header,
 		Payload:  &pb.Payload{Data: &pb.Payload_Payload{payload}},
 	}
 
-	prot.processMessage(msg)
+	prot.processMessage(prot.localNodePubkey, msg) // todo: pass sender
```
Comment (zalmen) [inline p2p/gossip/protocol.go]: Why the "todo"?

File: p2p/gossip/protocol.go [hunk 7]
```diff
@@ -277,37 +236,26 @@ func (prot *Protocol) processMessage(msg *pb.ProtocolMessage) {
...
-		if validity == Valid {
-			prot.propagateMessage(data.Bytes(), h, protocol)
-		} else {
-			metrics.InvalidGossipMessages.With(metrics.ProtocolLabel, protocol).Add(1)
-			// if the message is invalid we don't want to propagate it and we can return. If the message's validity is unknown,
-			// since the message is marked as old we can assume that there is another context that currently process this
-			// message and will determine its validity, therefore we can return in such case as well
-			return
-		}
+		// todo : - maybe block this peer since he sends us old messages
```
Comment (zalmen) [inline p2p/gossip/protocol.go:236]: add log print

**Review by zalmen — CHANGES_REQUESTED** [review]

**Review by zalmen — APPROVED** [review]

**Pull request merged**